### ISSUE TITLE
The "episode count" displayed against a podcast is inaccurate

### ISSUE THREAD

**Issue by boc-the-git** [madeofpendletonwool/PinePods#228]
One of my podcasts displays as `Episode Count: 50`.. however I can count the podcasts on the screen and it's more than 50.
The RSS feed for the podcast lists the most recent 50 episodes, and my assumption is that's what's being referenced in some way to determine the "episode count".

**Comment by madeofpendletonwool** [comment]
Really? That's one I wouldn't have expected. The episode count comes from the data the api returns server side. The count of episodes is pulled from that. So it's pretty certain the issue is due to the count from server side. Did the server actually pull all the episodes into the db is the next question then. Is it a public podcast? I could check that one out

**Comment by boc-the-git** [comment]
> Is it a public podcast?

It's not, it's that private podcast I've referenced. I can guarantee there are 50 podcasts on the RSS feed at all times, despite them having a backlog I assume of several thousand.

Episodes that are no longer on the RSS feed are remaining displayed in Pinepods (makes sense to me).

Is there any particular thing in the DB you'd like me to source? (probably won't be tonight)

**Comment by madeofpendletonwool** [comment]
Hmm... So are the owners of the feed removing episodes from the feed as they get stale? The current logic that Pinepods follows is this:

You add the feed in, Pinepods adds a reference to all the episodes in the Episodes table for quick reference later. From there, Pinepods displays the episode count based on the amount of episodes it was able to get from that feed. As Pinepods re-parses the feed for updates it will add new episodes on and not remove old episodes, as you've discovered, even if they're removed from the feed. 

So, if they have a backlog of several thousand but only so many are on the feed it would make sense Pinepods would only see the 50 it gathered from the feed. 

I'm not entirely certain how we would handle this in a better way, just because I don't know that we could parse for a backlog of episodes if they aren't part of the feed. Perhaps I'm misunderstanding something?

**Comment by boc-the-git** [comment]
> Hmm... So are the owners of the feed are removing episodes from the feed as they get stale? The current logic that Pinepods follows is this:
> 
> You add the feed in, Pinepods adds a reference to all the episodes in the Episodes table for quick reference later. From there, Pinepods displays the episode count based on the amount of episodes it was able to get from that feed. As Pinepods re-parses the feed for updates it will add new episodes on and not remove old episodes, as you've discovered, even if they're removed from the feed. 
> 
> So, if they have a backlog of several thousand but only so many are on the feed it would make sense Pinepods would only see the 50 it gathered from the feed. 
> 
> I'm not entirely certain how we would handle this in a better way, just because I don't know that we could parse for a backlog of episodes if they aren't part of the feed. Perhaps I'm misunderstanding something?

I agree you can't parse historical episodes, that's a limitation of the RSS feed for sure. 

It sounds to me like pinepods is showing "number of episodes that are on the RSS feed"? That doesn't make sense to me under any circumstance really. 

Somewhere in pinepods it's storing all the episodes it takes from the feed (currently 60+ for this podcast), and the count of those is the number that makes sense to show. 

I'd further say it makes sense to be showing the "unlistened count", but that's a separate issue. 

(I appreciate an RSS feed being incomplete like this is an edge case, but I don't think the current logic makes sense under any circumstances)

**Comment by madeofpendletonwool** [comment]
Sorry, I think I explained that poorly. 
What I mean to say is that when displayed on the screen here:
![image](https://github.com/user-attachments/assets/3d72973c-d0a3-404c-9c89-3cb7c05c2ba2)
That number of episodes is pulled from the database, not the feed. We don't reparse the feed to display those counts. The number is just pulled from the db. So that number represents the number of episodes the db has picked up for that particular podcast. Not the number of  episodes in the RSS feed. 

So I think we're on the same page. The logic represents the number of that show in the database. The unlistened count number is a great idea though, and would be really easy to add.

**Comment by madeofpendletonwool** [comment]
> Somewhere in pinepods it's storing all the episodes it takes from the feed (currently 60+ for this podcast), and the count of those is the number that makes sense to show.

You are right about this too. There's an episodes table. And we grab the count of episodes from the episode id for that podcast

**Comment by boc-the-git** [comment]
Ok, so that all makes sense from the point of view that you're describing how I also think it should behave, but I'm quite certain that isn't happening in my instance.
I'll try to get connected to the db to pull info and confirm.

**Comment by boc-the-git** [comment]
@madeofpendletonwool, checking in the DB, there appears to be a static `episodecount` field on the Podcasts table (that I expect is "the number of the RSS feed"), and it _does not match up_ with how many records exist in the Episodes table for that specific podcast. Confirmed with 3 different podcasts in my DB.

Perhaps that `episodecount` field is meant to be dynamic or get recalculated, or something?

**Comment by boc-the-git** [comment]
> Hmm... So are the owners of the feed are removing episodes from the feed as they get stale?

Now that I'm aware of this as a thing and paying attention to it, I can actually see at least 3 podcasts of 22 in my server have similar behaviour.. small RSS feeds despite having put out hundreds/thousands of episodes over the years. So it's not an exceptional scenario.

(of course, given where our discussion has got to, I now think there's a fundamental flaw in the count, as opposed to this being specifically related to a reduced RSS feed.. just writing this to highlight it's not "one podcast")

**Comment by madeofpendletonwool** [comment]
Well what do you know, I guess after working on this thing long enough I start to forget how certain parts work. You're right. 

`````
def add_podcast(cnx, database_type, podcast_values, user_id):
    cursor = cnx.cursor()
    print(f"Podcast values '{podcast_values}'")

    try:
        # Check if the podcast already exists for the user
        if database_type == "postgresql":
            query = 'SELECT PodcastID FROM "Podcasts" WHERE FeedURL = %s AND UserID = %s'
        else:  # MySQL or MariaDB
            query = "SELECT PodcastID FROM Podcasts WHERE FeedURL = %s AND UserID = %s"

        cursor.execute(query, (podcast_values['pod_feed_url'], user_id))
        result = cursor.fetchone()
        print(f"Result: {result}")
        print("Checked for existing podcast")

        if result is not None:
            # Podcast already exists for the user, return False
            cursor.close()
            return False

        # Extract category names and convert to comma-separated string
        categories = podcast_values['categories']
        print(f"Categories: {categories}")

        if isinstance(categories, dict):
            category_list = ', '.join(categories.values())
        elif isinstance(categories, list):
            category_list = ', '.join(categories)
        elif isinstance(categories, str):
            category_list = categories
        else:
            category_list = ''

        # Insert the podcast into the database
        if database_type == "postgresql":
            add_podcast_query = """
                INSERT INTO "Podcasts"
                (PodcastName, ArtworkURL, Author, Categories, Description, EpisodeCount, FeedURL, WebsiteURL, Explicit, UserID)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s) RETURNING PodcastID
            """
            explicit = podcast_values['pod_explicit']
        else:  # MySQL or MariaDB
            add_podcast_query = """
                INSERT INTO Podcasts
                (PodcastName, ArtworkURL, Author, Categories, Description, EpisodeCount, FeedURL, WebsiteURL, Explicit, UserID)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            explicit = 1 if podcast_values['pod_explicit'] else 0
```


It just gets the count once. Yeah that field needs to be updated on subsequent refreshes. Just flat out forgot to implement that. Whoops. I definitely just validated what you're saying. The number is certainly less in most instances on the podcasts in the db count vs the count shown on the podcast itself. Well, easy enough fix! Just needs a quick sql function to update that field on refreshing.

**Comment by boc-the-git** [comment]
The key takeaway from this, is I finally got around to installing a DB client on my personal laptop.. long overdue!

**Comment by madeofpendletonwool** [comment]
Glad I could get you past the breaking point of needing one. Everyone needs a little more sql in their life lol

This is overkill on the investigation at this point but here's where we initially get the number of episodes that goes into the count. It would be accurate (at time of adding the podcast initially anyway) due to it being a len count of the number of entries. Which is what it pulls the data from to actually add to the database anyway. So yeah, definitely just need something to update it on refresh. 

`
def get_podcast_values(feed_url, user_id):
    import feedparser
    import json
    # Parse the feed
    d = feedparser.parse(feed_url)

    # Initialize podcast_values as a dictionary
    podcast_values = {
        'pod_title': d.feed.title if hasattr(d.feed, 'title') else None,
        'pod_artwork': d.feed.image.href if hasattr(d.feed, 'image') and hasattr(d.feed.image, 'href') else None,
        'pod_author': d.feed.author if hasattr(d.feed, 'author') else None,
        'categories': [],
        'pod_description': d.feed.description if hasattr(d.feed, 'description') else None,
        'pod_episode_count': len(d.entries),
        'pod_feed_url': feed_url,
        'pod_website': d.feed.link if hasattr(d.feed, 'link') else None,
        'pod_explicit': False,
        'user_id': user_id
    }
`

**Comment by madeofpendletonwool** [comment]
Fix implemented as discussed #243 

Should be good to go moving forward! This along with numerous other fixes will be added into a release this weekend. 0.6.4
